plugin/call_hierarchy.py

```diff
@@ -0,0 +1,157 @@
+    ) -> None:
+        self.window = window
+        self.session_name = session_name
+        self.direction = direction
+        self.root_elements = root_elements
+
+    def get_children(self, element: Optional[CallHierarchyItem]) -> Promise[List[CallHierarchyItem]]:
+        if element is None:
+            return Promise.resolve(self.root_elements)
+        wm = windows.lookup(self.window)
+        if not wm:
+            return Promise.resolve([])
+        for session in wm.get_sessions():
+            if not session.has_capability('callHierarchyProvider'):
+                continue
+            if session.config.name == self.session_name:
+                break
+        else:
+            return Promise.resolve([])
+        if self.direction == CallHierarchyDirection.IncomingCalls:
+            params = cast(CallHierarchyIncomingCallsParams, {'item': element})
+            return session.send_request_task(Request.incomingCalls(params)).then(self._handle_incoming_calls_async)
+        elif self.direction == CallHierarchyDirection.OutgoingCalls:
+            params = cast(CallHierarchyOutgoingCallsParams, {'item': element})
+            return session.send_request_task(Request.outgoingCalls(params)).then(self._handle_outgoing_calls_async)
+        return Promise.resolve([])
+
+    def get_tree_item(self, element: CallHierarchyItem) -> TreeItem:
+        command_url = sublime.command_url('lsp_open_location', {
+            'location': {
+                'targetUri': element['uri'],
+                'targetRange': element['range'],
+                'targetSelectionRange': element['selectionRange']
+            },
+            'session_name': self.session_name
+        })
+        return TreeItem(
+            element['name'],
+            kind=SYMBOL_KINDS.get(element['kind'], sublime.KIND_AMBIGUOUS),
+            description=element.get('detail', ""),
+            tooltip="{}:{}".format(parse_uri(element['uri'])[1], element['selectionRange']['start']['line'] + 1),
+            command_url=command_url
+        )
+
+    def _handle_incoming_calls_async(
+        self, response: Optional[List[CallHierarchyIncomingCall]]
+    ) -> List[CallHierarchyItem]:
+        return [incoming_call['from'] for incoming_call in response] if response else []
+
+    def _handle_outgoing_calls_async(
+        self, response: Optional[List[CallHierarchyOutgoingCall]]
+    ) -> List[CallHierarchyItem]:
+        return [outgoing_call['to'] for outgoing_call in response] if response else []
+
+
+class LspCallHierarchyCommand(LspTextCommand):
+
+    capability = 'callHierarchyProvider'
+
+    def is_visible(self, event: Optional[dict] = None, point: Optional[int] = None) -> bool:
+        if self.applies_to_context_menu(event):
+            return self.is_enabled(event, point)
+        return True
+
+    def run(self, edit: sublime.Edit, event: Optional[dict] = None, point: Optional[int] = None) -> None:
+        self._window = self.view.window()
```

Comment:
Would generally prefer validating with window manager at every point so that we stop processing everything as soon as it goes away.

Reply:
Is this comment still relevant? Because the CallHierarchyDataProvider checks each time in `get_children` with the window manager for the session with the correct name, and if the session dies you can't interact with the call hierarchy sheet anymore. So as I see it, the only place where we could add another check would be in `_handle_response_async` here, but the time between initial request and response is probably only a few milliseconds.

Otherwise, could you tell what exactly should I change?